SwrveSDK/src/main/java/com/swrve/sdk/SwrveImp.java

```diff
@@ -970,36 +976,36 @@ private boolean supportsDeviceFilter(String requirement) {
     }
 
     protected void downloadAssets(final Set<String> assetsQueue) {
-        assetsCurrentlyDownloading = true;
+        assetManager.setAssetsCurrentlyDownloading(true);
         ExecutorService resourceDownloadExecutor = Executors.newSingleThreadExecutor();
         resourceDownloadExecutor.execute(SwrveRunnables.withoutExceptions(new Runnable() {
             @Override
             public void run() {
                 Set<String> assetsToDownload = filterExistingFiles(assetsQueue);
                 for (String asset : assetsToDownload) {
+
                     boolean success = downloadAssetSynchronously(asset);
                     if (success) {
-                        synchronized (assetsOnDisk) {
-                            assetsOnDisk.add(asset);
+                        if (success) {
+                            assetManager.addAsset(asset);
+                        } else {
+                            assetManager.addFailedAsset(asset);
                         }
                     }
+
+                    if (assetManager.hasFailedAssets()) {
+                        downloadAssets(assetManager.getFailedAssets());
```

Comment:
Yeah this is where my logic skills were concerned. In theory this would run recursively until the stack ran out of space so a finite depth level would be essential IMO